src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+    lo_frequency: int = 0
+    lo_power: int = 0
+    gain: int = 0
+    attenuation: int = 0
+    power_range: int = 0
+    filters: Optional[dict] = None
+
+
+class PulseSimulator(Controller):
+    """Runs quantum dynamics simulation of model of device.
+
+    Interfaces with Qibolab. Useful for testing code without requiring
+    access to hardware.
+    """
+
+    PortType = DummyPort  # Placeholder. Copied over from dummy, may not be required.
+
+    def __init__(
+        self,
+        simulation_config: dict,
+        model_config: dict,
+        sim_opts: Optional[None] = None,
+    ):
+        """Set emulator configuration.
+
+        Args:
+            simulation_config (dict): Simulation configuration dictionary.
+            model_config (dict): Model configuration dictionary.
+            sim_opts (optional): Simulation backend specific object specifying simulation options.
+        """
+        self.simulation_config = simulation_config
+        self.model_config = model_config
+        self.sim_opts = sim_opts
+
+        self.all_simulation_backends = {"Qutip": QutipSimulator}
+
+        self.available_sweep_parameters = {
+            Parameter.amplitude,
+            Parameter.duration,
+            Parameter.frequency,
+            Parameter.relative_phase,
+            Parameter.start,
+        }
+
+        self.update()
+
+    @property
+    def sampling_rate(self):
+        return self.model_config["sampling_rate"]
+
+    def update(self):
+        """Updates the pulse simulator by loading all parameters from
+        `self.model_config` and `self.simulation_config`."""
+        self.simulation_backend_name = self.simulation_config["simulation_backend_name"]
+        self.device_name = self.model_config["device_name"]
+        self.model_name = self.model_config["model_name"]
+        self.emulator_name = f"{self.device_name} emulator running {self.model_name} on {self.simulation_backend_name} backend"
+        self.simulation_backend = self.all_simulation_backends[
+            self.simulation_backend_name
+        ](self.model_config, self.sim_opts)
+
+        self.platform2simulator_channels = self.model_config[
+            "platform2simulator_channels"
+        ]
+
+        self.sim_sampling_boost = self.simulation_config["sim_sampling_boost"]
+        self.instant_measurement = self.simulation_config["instant_measurement"]
+        self.runcard_duration_in_dt_units = self.model_config[
+            "runcard_duration_in_dt_units"
+        ]
+        # self.readout_error = self.model_config["readout_error"]
+        self.readout_error = {}
+        for k, v in self.model_config["readout_error"].items():
+            self.readout_error.update({int(k): v})
+        self.exec_params = ExecutionParameters
+        self.default_nshots = self.simulation_config["default_nshots"]
+        self.exec_params.nshots = self.default_nshots
+        self.exec_params.acquisition_type = AcquisitionType.DISCRIMINATION
```

Comment:
I guess that should be
```suggestion
        self.default_nshots = self.simulation_config["default_nshots"]
        self.exec_params = ExecutionParameters(
            nshots=self.default_nshots,
            acquisition_type=AcquisitionType.DISCRIMINATION
        )
```
however I don't think you need this attribute at all, as the `Platform` already handles default execution parameters and will pass them to `Instrument.play`.

Also, adding new attributes to an object outside `__init__` is usually not very good practice as it makes it harder to understand classes ([pylint would have complained](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/attribute-defined-outside-init.html) if we were not ignoring warnings).

Reply:
Moreover, the `ExecutionParameters` are the parameters of the execution (quite explicit...), and they are not meant to be stored by the controller object, but they are passed to the `.play()` method, and they may change at every call.
https://github.com/qiboteam/qibolab/blob/76def9337aa3e60fcc35027faa584c9ed41458d5/src/qibolab/platform.py#L178-L180

So, please, do not add them during initialization, and do not store them at all in the `PulseSimulator` instance.

Reply:
Removed in 170b78df1260f4f8dad1b400c63a0fa59a3227f0